### PR TITLE
Explicit Update with Indices implementation

### PR DESCRIPTION
**Pull request by Bensuo**
Implements spec PR - https://github.com/intel/llvm/pull/12486

- Implementation of explicit update using indices
- E2E tests for each update type: accessor/pointer/scalar/nd_range
- Tests for errors and other edge cases

### CODE REVIEW

File: sycl/test-e2e/Graph/Explicit/update_with_indices_scalar.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// Extra run to check for leaks in Level Zero using UR_L0_LEAKS_DEBUG
+// RUN: %if level_zero %{env UR_L0_LEAKS_DEBUG=1 %{run} %t.out 2>&1 | FileCheck %s %}
+//
+// CHECK-NOT: LEAK
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/update_with_indices_scalar.cpp]: Use ` --implicit-check-not=LEAK ` instead

File: sycl/test-e2e/Graph/Explicit/update_with_indices_scalar.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+// Unsupported: opencl, level_zero
+
+// Tests updating a graph node scalar argument using index-based explicit update
+
+#include "../graph_common.hpp"
+
+int main() {
+  queue Queue{{sycl::ext::intel::property::queue::no_immediate_command_list{}}};
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/update_with_indices_scalar.cpp]: Can remove `no_immediate_command_list` property and replace with `SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS` in run line

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 3]
```diff
@@ -301,12 +319,27 @@ class __SYCL_EXPORT executable_command_graph {
...
+  }
+
 protected:
   /// Constructor used by internal runtime.
   /// @param Graph Detail implementation class to construct with.
   /// @param Ctx Context to use for graph.
   executable_command_graph(const std::shared_ptr<detail::graph_impl> &Graph,
-                           const sycl::context &Ctx);
+                           const sycl::context &Ctx,
+                           const property_list &PropList = {});
```
Comment (EwanC) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp:377]: doxy comment

File: sycl/include/sycl/detail/pi.h
Comment (EwanC) [inline sycl/include/sycl/detail/pi.h:1]: There's probably some ABI tests that need updated with the new PI symbols

File: sycl/plugins/unified_runtime/CMakeLists.txt [hunk 4]
```diff
@@ -56,15 +56,9 @@ endif()
...
-  # commit 47102cb2b275472054803a9399b5b977ef210b23
-  # Merge: 186bfb9d 2390664d
-  # Author: aarongreig <aaron.greig@codeplay.com>
-  # Date:   Fri Feb 9 12:21:47 2024 +0000
-  #     Merge pull request #1328 from Bensuo/maxime/events-reset-bugfix
-  #     [EXP][CMDBUF] Reset events for multiple submissions (bugfix)
-  set(UNIFIED_RUNTIME_TAG 47102cb2b275472054803a9399b5b977ef210b23)
-  
+  set(UNIFIED_RUNTIME_REPO "https://github.com/bensuo/unified-runtime.git")
+  set(UNIFIED_RUNTIME_TAG ewan/update)
```
Comment (EwanC) [inline sycl/plugins/unified_runtime/CMakeLists.txt]: Whatever the tip sycl branch is using should be fine here now that `ewan/update` is merged

File: sycl/source/detail/handler_impl.hpp [hunk 5]
```diff
@@ -117,6 +120,11 @@ class handler_impl {
 
   // Extra information for semaphore interoperability
   sycl::detail::pi::PiInteropSemaphoreHandle MInteropSemaphoreHandle;
+
+  // Storage any dynamic parameters which have been flagged for registration in
```
Comment (EwanC) [inline sycl/source/detail/handler_impl.hpp]: Would maybe put "SYCL Graph dynamic parameters", so that a developer not familiar with the dynamic parameter terminology will know where to look.

**Review by EwanC — COMMENTED** [review]
We should be checking somewhere if the backend device supports update, reported by `UR_DEVICE_INFO_COMMAND_BUFFER_UPDATE_SUPPORT_EXP`, and error if that's not available and a user tries to update

File: sycl/include/sycl/detail/pi.h
Comment (Bensuo) [reply sycl/include/sycl/detail/pi.h:1]: These have been updated now.

File: sycl/plugins/unified_runtime/CMakeLists.txt (hunk 4 above)
Comment (Bensuo) [reply sycl/plugins/unified_runtime/CMakeLists.txt]: It is now pointing at tip.

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 6]
```diff
@@ -350,13 +384,60 @@ class command_graph : public detail::modifiable_command_graph {
...
 };
 
+namespace detail {
+class __SYCL_EXPORT dynamic_parameter_base {
+public:
+  dynamic_parameter_base(
+      sycl::ext::oneapi::experimental::command_graph<graph_state::modifiable>
+          Graph);
+
+  void register_with_node(int ArgIndex, handler &CGH);
```
Comment (EwanC) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp]: in our spec draft this is just called `register()` rather than `register_with_node()`

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 7]
```diff
@@ -350,13 +384,60 @@ class command_graph : public detail::modifiable_command_graph {
...
+                       ? sycl::detail::kernel_param_kind_t::kind_pointer
+                       : sycl::detail::kernel_param_kind_t::kind_std_layout;
+
+public:
+  dynamic_parameter(experimental::command_graph<graph_state::modifiable> Graph)
+      : detail::dynamic_parameter_base(Graph), MValue() {}
+
+  dynamic_parameter(ValueT InitialValue,
+                    experimental::command_graph<graph_state::modifiable> Graph)
+      : detail::dynamic_parameter_base(Graph), MValue(InitialValue) {}
```
Comment (EwanC) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp:460]: This constructor isn't in the current spec draft, and I know the advantage is that a user doesn't need to pass the template type to the constructor.
e.g.
```
exp_exp::dynamic_parameter(V, G);
````
rather than
```
exp_exp::dynamic_parameter<decltype(V)> (G);
```
We should decide to add this to the spec or not.

Comment (Bensuo) [reply sycl/include/sycl/ext/oneapi/experimental/graph.hpp:460]: I just pushed this to the spec PR, but can revert if there are objections to it. It might even make more sense to make it the only constructor, otherwise we probably need to define the behaviour of a `dynamic_parameter` which has never had it's value set by `update()`.

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp (hunk 6 above)
Comment (Bensuo) [reply sycl/include/sycl/ext/oneapi/experimental/graph.hpp]: I hadn't updated this in the spec yet but I ran into conflicts with the `register` keyword in C++ that I don't believe are possible to work around, so I renamed the function instead.

File: sycl/test-e2e/Graph/Explicit/update_with_indices_accessor.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      Queue.get_context(),
+      Queue.get_device(),
+      {exp_ext::property::graph::assume_buffer_outlives_graph{}}};
+  std::vector<int> HostDataA(N, 0);
+  std::vector<int> HostDataB(N, 0);
+
+  buffer BufA{HostDataA};
+  buffer BufB{HostDataB};
+  BufA.set_final_data(nullptr);
+  BufB.set_final_data(nullptr);
```
Comment (mfrancepillois) [inline sycl/test-e2e/Graph/Explicit/update_with_indices_accessor.cpp]: Why do you use `set_final_data` instead of `set_write_back` as in other tests?
Could you please add a comment to explain this?

File: sycl/test-e2e/Graph/Explicit/update_with_indices_accessor.cpp [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+  exp_ext::dynamic_parameter InputParam(BufA.get_access(), Graph);
+
+  auto KernelNode = Graph.add([&](handler &cgh) {
+    // Register the input pointer, we should be using set_arg but can't
+    // currently test that with CUDA
+    // cgh.set_arg(0, ptrA)
```
Comment (mfrancepillois) [inline sycl/test-e2e/Graph/Explicit/update_with_indices_accessor.cpp]:
```suggestion
    // cgh.set_arg(0, BufA)
```
I don't think this comment is accurate.

**Review by mfrancepillois — APPROVED** [review]
Nice PR!
I think you should rebase it on the latest version of the sycl branch (to avoid confusion with regard to nodes and their copies in exec_graph (Commit https://github.com/intel/llvm/commit/5337a8aa67a2089edffbc092bcdceab03eed1001#diff-660b1c0efae10a9e21cba6738e34672b1a922f53eb2d466d8e0908d3512f042a)).
But then we will need to use `%{l0_leak_check}` instead of `UR_L0_LEAKS_DEBUG=1` in your tests.

File: sycl/test-e2e/Graph/Explicit/update_nd_range.cpp
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/update_nd_range.cpp:1]: I know you're going to add more tests anyway, but an outline of tests I'd expect to see:
* [ ] Dynamic parameter registered with more than one node.
* [ ] Check for an error if you try to register a dynamic_parameter on a node not in a graph used on creation of dynamic_parameter.
* [ ] A single node registered with more than one dynamic_parameter registered and upated.
* [ ] Update twice on same dynamic_parameter, then update node in exec-graph, then submit and check
   only most recent value is preserved.
* [ ] Graph update() on node not in graph throws an error.
* [ ] Graph update on node that hasn't been updated with a dynamic_parameter matches spec.
* [ ] Error on update when updatable property is not set.
* [ ] Create more than one executable graph from a modifiable graph with dynamic parameter registered, and check updating one of the graphs doesn't affect the other.

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp (hunk 6 above)
Comment (EwanC) [reply sycl/include/sycl/ext/oneapi/experimental/graph.hpp]: hadn't considered that, but a very good point. `register_with_node` sounds fine to me

File: sycl/source/detail/graph_impl.cpp [hunk 10]
```diff
@@ -912,6 +929,205 @@ exec_graph_impl::enqueue(const std::shared_ptr<sycl::detail::queue_impl> &Queue,
...
+  UpdateDesc.exec_info_list = ExecInfoDescs.data();
+
+  UpdateDesc.global_work_offset = &NDRDesc.GlobalOffset[0];
+  UpdateDesc.global_work_size = &NDRDesc.GlobalSize[0];
+  UpdateDesc.local_work_size = LocalSize;
+  UpdateDesc.num_work_dim = NDRDesc.Dims;
+
+  auto Command = MCommandMap[Node];
+  pi_result Res = Plugin->call_nocheck<
+      sycl::detail::PiApiKind::piextCommandBufferUpdateKernelLaunch>(
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.cpp:1385]: General comment that we should put something in the design doc about update, even if it's just adding the UR entry-point to the table of UR entry-points we have, but ideally any other interesting details (probably around buffer accessors since that seemed like the tricky bit)

File: sycl/test-e2e/Graph/Explicit/update_with_indices_accessor.cpp (hunk 8 above)
Comment (Bensuo) [reply sycl/test-e2e/Graph/Explicit/update_with_indices_accessor.cpp]: No reason, I think that was just the first one that came to mind when I wrote it. I've changed it to `set_write_back` anyway just for consistency.

File: sycl/test-e2e/Graph/Explicit/update_with_indices_accessor.cpp (hunk 9 above)
Comment (Bensuo) [reply sycl/test-e2e/Graph/Explicit/update_with_indices_accessor.cpp]: Good catch, though I think it should be `Acc` in the comment so I've updated it as such.

File: sycl/test-e2e/Graph/Explicit/update_nd_range.cpp
Comment (EwanC) [reply sycl/test-e2e/Graph/Explicit/update_nd_range.cpp:1]: Trying to think of a test with sub-graphs too, where the dynamic_parameter is registered on a node in the child graph. Something like:
```cpp
 exp_ext::dynamic_parameter<int *> InputParam(Graph);
 auto KernelNode = Graph.add([&](handler &cgh) {
    InputParam.register_with_node(0, cgh);
    cgh.single_task([=]() {
      for (size_t i = 0; i < N; i++) {
        ptr[i] = i;
      }
    });
  });

  auto ExecGraph = Graph.finalize(exp_ext::property::graph::updateable{});

 auto SubGraphNode = ParentGraph.add([&](handler &cgh) {
    cgh.ext_oneapi_graph(ExecGraph);
  });
  auto ExecParentGraph = ParentGraph.finalize(exp_ext::property::graph::updateable{});

  InputParam.update(foo);
  ExecGraph.update(KernelNode);
  ExecParentGraph.update(KernelNode); // Should be an error to try and update.

  Queue.ext_oneapi_graph(ExecGraph).wait(); // Should have been updated to foo
  Queue.ext_oneapi_graph(ExecParentGraph ).wait(); // Should not have been updated
```

**Comment by Bensuo** [comment]
> We should be checking somewhere if the backend device supports update, reported by `UR_DEVICE_INFO_COMMAND_BUFFER_UPDATE_SUPPORT_EXP`, and error if that's not available and a user tries to update
@EwanC 

Do we want to also add this as an aspect or `device::get_info()` query for the user to use? Otherwise they can't actually check if the device supports updating before attempting it. We used `device::get_info()` before for general graph support though I see similar device info queries mapping to aspects instead.

**Comment by EwanC** [comment]
> Do we want to also add this as an aspect or `device::get_info()` query for the user to use? Otherwise they can't actually check if the device supports updating before attempting it. We used `device::get_info()` before for general graph support though I see similar device info queries mapping to aspects instead.

Good question, so the only DPC++ backend that will get up not having guaranteed support is OpenCL (CUDA, HIP, L0) all will in the near future. Then what I think we need to decide is if it is worth emulating the update for this target to avoid the user having to check a device capability, in which case throwing an exception until that work is done is fine. Also once we pick a aspect/`device::get_info()` path it will set a precedent for other things, e.g. graph owned memory allocations, profiling at the event level, external event sync so that's something to consider in terms of which one we go for.

File: sycl-fusion/jit-compiler/lib/fusion/FusionHelper.cpp [hunk 11]
```diff
@@ -64,6 +64,7 @@ Expected<std::unique_ptr<Module>> helper::FusionHelper::addFusedKernel(
     // correct name and signature only during the fusion pass.
     auto *F = Function::Create(FT, GlobalValue::LinkageTypes::ExternalLinkage,
                                "fused_kernel", *NewMod);
+    F->IsNewDbgInfoFormat = UseNewDbgInfoFormat;
```
Comment (EwanC) [inline sycl-fusion/jit-compiler/lib/fusion/FusionHelper.cpp]: Seeing a kernel fusion commit in this git branch which i don't think should be here

Comment (Bensuo) [reply sycl-fusion/jit-compiler/lib/fusion/FusionHelper.cpp]: This should be gone now

File: sycl/test-e2e/Graph/Explicit/update_nd_range.cpp [hunk 12]
```diff
@@ -0,0 +1,59 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// Extra run to check for leaks in Level Zero using UR_L0_LEAKS_DEBUG
+// RUN: %if level_zero %{env UR_L0_LEAKS_DEBUG=1 %{run} %t.out 2>&1 | FileCheck %s %}
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/update_nd_range.cpp]:
```cpp
// RUN: %if level_zero %{env SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=0 %{l0_leak_check} %{run} %t.out 2>&1 | FileCheck %s --implicit-check-not=LEAK %}
// Extra run to check for immediate-command-list in Level Zero
// RUN: %if level_zero && linux %{env SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=1 %{l0_leak_check} %{run} %t.out 2>&1 | FileCheck %s --implicit-check-not=LEAK %}
```

File: sycl/test-e2e/Graph/Explicit/update_nd_range.cpp [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+//
+// CHECK-NOT: LEAK
+// Unsupported: opencl, level_zero
+
+// Tests updating a graph node using index-based explicit update
+
+#include "../graph_common.hpp"
+
+int main() {
+  queue Queue{{sycl::ext::intel::property::queue::no_immediate_command_list{}}};
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/update_nd_range.cpp]:
```suggestion
  queue Queue
```

File: sycl/test-e2e/Graph/Explicit/update_with_indices_multiple_exec_graphs.cpp [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  if (!are_graphs_supported(Queue)) {
+    return 0;
+  }
+
+  const size_t N = 1024;
+
+  exp_ext::command_graph Graph{Queue.get_context(), Queue.get_device()};
+
+  int *ptrA = malloc_device<int>(N, Queue);
+  int *ptrB = malloc_device<int>(N, Queue);
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/update_with_indices_multiple_exec_graphs.cpp]:
```suggestion
  int *PtrA = malloc_device<int>(N, Queue);
  int *PtrB = malloc_device<int>(N, Queue);
```

File: sycl/test-e2e/Graph/Explicit/update_with_indices_ordering.cpp [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  // Use a large N to try and make the kernel slow
+  const size_t N = 1 << 16;
+  // Loop inside kernel to make even slower (too large N runs out of memory)
+  const size_t NumKernelLoops = 4;
+  const size_t NumSubmitLoops = 8;
+
+  exp_ext::command_graph Graph{Queue.get_context(), Queue.get_device()};
+
+  int *ptrA = malloc_device<int>(N, Queue);
+  int *ptrB = malloc_device<int>(N, Queue);
```
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/update_with_indices_ordering.cpp]:
```suggestion
  int *PtrA = malloc_device<int>(N, Queue);
  int *PtrB = malloc_device<int>(N, Queue);
```

Same for other tests

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 16]
```diff
@@ -113,6 +119,10 @@ class __SYCL_EXPORT node {
   /// submission.
   static node get_node_from_event(event nodeEvent);
 
+  /// Update the ND-Range of this node if it is a kernel execution node
+  template <int Dimensions>
+  void update_nd_range(nd_range<Dimensions> executionRange);
+
```
Comment (EwanC) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp:125]: Need an `update_range` entry-point too now with test

File: sycl/source/detail/graph_impl.cpp [hunk 17]
```diff
@@ -736,6 +757,35 @@ void exec_graph_impl::createCommandBuffers(
...
+                                 const property_list &PropList)
+    : MSchedule(), MGraphImpl(GraphImpl), MPiSyncPoints(), MContext(Context),
+      MRequirements(), MExecutionEvents(),
+      MIsUpdateable(PropList.has_property<property::graph::updateable>()) {
+
+  // If the graph has been marked as updateable then check if the backend
+  // actually supports that.
+  if (MIsUpdateable) {
+    pi_bool SupportsUpdate = PI_FALSE;
+    bool call_successful =
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.cpp]:
```suggestion
    bool CallSuccessful =
```

File: sycl/source/detail/graph_impl.cpp [hunk 18]
```diff
@@ -1091,6 +1149,252 @@ void exec_graph_impl::duplicateNodes() {
...
+  auto &ExecCG =
+      *(static_cast<sycl::detail::CGExecKernel *>(Node->MCommandGroup.get()));
+  // Copy args because we may modify them
+  std::vector<sycl::detail::ArgDesc> NodeArgs = ExecCG.getArguments();
+  // Copy NDR desc since we need to modify it
+  auto NDRDesc = ExecCG.MNDRDesc;
+
+  // TODO: Try and refactor this stuff because it is copied from
+  // enqueueImpCommandBufferKernel
+  // TODO: Probably have to handle eliminated args too
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.cpp]: These TODOs intended for this PR? If not can you create an internal ticket so we don't forget this improvement

File: sycl/source/detail/graph_impl.hpp [hunk 19]
```diff
@@ -75,6 +76,10 @@ inline node_type getNodeTypeFromCG(sycl::detail::CG::CGTYPE CGType) {
 /// Implementation of node class from SYCL_EXT_ONEAPI_GRAPH.
 class node_impl {
 public:
+  using id_type = uint64_t;
+
+  // Unique identifier for this node.
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp]:
```suggestion
  /// Unique identifier for this node.
```

**Review by EwanC — APPROVED** [review]
LGTM, nice work

File: sycl/test-e2e/Graph/Explicit/update_nd_range.cpp
Comment (EwanC) [inline sycl/test-e2e/Graph/Explicit/update_nd_range.cpp:1]: Could we also have a test like Greg Lueck's question from the spec PR
```c++
sycl_ext::command_graph myGraph(myContext, myDevice);  // Creates a modifiable graph

int myScalar = 42;
dynamic_parameter dynParamScalar(myGraph, myScalar);

node nodeA = myGraph.add([&](handler& cgh) {
  dynParamScalar.register_with_node(cgh, 0);
  cgh.parallel_for(range {n}, myKernel);
});

dynParamScalar.update(1024);  // Change the value of the parameter before the graph is finalized

auto execGraph = myGraph.finalize();
```
and check we do get 1024

**Comment by Bensuo** [comment]
Closing in favour of upstream PR: https://github.com/intel/llvm/pull/12840
